### ISSUE TITLE
Handling output from the Neural network, when bidding

### ISSUE THREAD

**Issue by ThorvaldAagaard** [lorserker/ben#45]
I have looked at the rules for selecting a bid, and I think that adding some simple rules might improve the bidding a lot.

Take this hand
![image](https://github.com/lorserker/ben/assets/10560652/4920d359-dd04-4c46-a421-4fa672dfaaa9)

The Bot is in 4th position, and the bidding starts:
![image](https://github.com/lorserker/ben/assets/10560652/4f54b5d1-5bf4-45c7-9e22-559cc69ba6a1)

The result from the Neural network is
![image](https://github.com/lorserker/ben/assets/10560652/f24c90b6-937b-41b6-83e9-f8fe0df02249)

With the current threshold of 0.1 the bot will consider these two bids, and creates a simulation of how the bidding might continue based on op to 64 boards, that would match the bidding until now.

As the simulation is random, BEN sometimes bid 1S and other times 2S.

In my opinion any bid with a score from the neural network above 75% (a guess), should be selected without looking at other possible bids, as I expect the neural network to be much more precise than the simulation.

Just to test it I changed the threshold from 0.1 to 0.001, and now these 4 bids should be considered
![image](https://github.com/lorserker/ben/assets/10560652/ca330286-6b6b-4686-b06f-42f6569db503)

What is interesting is that the simulation now suggested that 3S was the best bid.

But lowering the threshold will also show the problem with the simulation, because North after this bidding
![image](https://github.com/lorserker/ben/assets/10560652/04a44cbd-699a-49d4-8b49-1d0001b90d99)
will evaluate these bids
![image](https://github.com/lorserker/ben/assets/10560652/77834806-339c-4119-9d05-d1195f04f73c)
and actually select 5D - a bid that is absolutely terrible.

So we could change the rule to be:

1; If Insta_score > 75% select that bid (Should be configurable)
2: Sort the bids in descending order of insta_score, and when 80% (Configurable) of possible bids are found use these as candidates.

The second option is to include a 3rd bid if the two bids with highest insta_score is below 80% as that could indicate that the neural network has no clear preference.

I can try playing around with these numbers to see if I can find the best numbers

What do you think about this approach?

**Comment by lorserker** [comment]
@ThorvaldAagaard this logic is quite tricky to find the optimal settings and it makes sense to tune it more. the way i select the bids currently is likely not the best as i have not tried hard to tune it.

to choose the settings objectively, it makes sense to run ben vs ben bidding matches and compare different versions of the bid selection logic. more details about how to run ben vs ben bidding are [here](https://github.com/lorserker/ben/tree/main/scripts/match/bidding)

**Comment by ThorvaldAagaard** [comment]
Yes, I will try the Ben vs Ben, and will probably move some of the configuration to the conf-file

In principle I want to have the same EW during all runs and see where the best score for NS is, and as long as it is the same Neural network I think it will be fine.

I will return when I have some results

**Comment by lorserker** [comment]
the model outputs a probability distribution over all possible bids.
if one of the bids is 98% then it means the model very strongly suggests it, and is also very confident that this bid is correct.
it's hard to tell what exactly makes the model confindent. usually it's many training examples where such bid was correct.
when a strongly suggested bid is obviously wrong and can't possibly come from training examples (e.g passing cuebids) then it could be due to either of these:
- the model mistakes the situation for another situation where it's correct to pass. this can be due to low capacity of the nn that it couldn't remember every unique situation and starts lumping together uncommon situations treating it the same. it could also be caused by the exact opposite problem, overfitting: say in the data there was a very similar hand where it was correct to pass, but the bidding was slightly different. the model is now triggered by the cards it holds and pays more attention to its own cards than to the bidding.
- bias: "pass is often the correct bid, so we might as well try it here :)"
- bugs

a lot of text without giving a clear answer, sorry.
the short version is: i have no idea what is going on

**Comment by ThorvaldAagaard** [comment]
I have been looking at another board, that could explain the problem:

With this hand BEN opens 1♠
![image](https://github.com/lorserker/ben/assets/10560652/0a30e744-f585-4ff3-94ac-05accfbbd40b)
and after partners support to 2♠ the neural network suggest 4♣. That is not the right bid, so lets look at the different options.

Lowering the threshold 3 different bids are possible
![image](https://github.com/lorserker/ben/assets/10560652/486e0108-065c-4c23-9f10-4013b9093d0d)

Now I asked GIB, what it would bid, and 2N is showing: Unspecified shortness, game try -- 5+ S; 17-18 total points
that fits perfectly to the hand

The problem is that 2N can be bid on hands with any shortage, but 4♣ shows short club, and even when missing some hcp, it seems to fit the description better.

The error is clear, when we look at the hands generated after the 4♣ bid, as most of the hands is much stronger than the actual hand
![image](https://github.com/lorserker/ben/assets/10560652/428a9565-1b24-40d4-919d-70746ce482db)

I have no idea about how to solve this, unless it is possible to learn the Neural Network that when matching hcp is more important than shape.

**Comment by ThorvaldAagaard** [comment]
Closing this for now, as we have no improvements planned, and no clear idea of what should be changed.
